Keyword `$vocabulary` (#230)
Intelligence-AI/learnjsonschema.com: content/2020-12/core/vocabulary.markdown

  - vocabulary: core
    keyword: $schema
---

> Vocabularies are sets of keywords and their meanings used to validate JSON documents. This keyword is crucial for meta-schemas, which are schemas that define other schemas.

Reviewer: I think the first sentence is a bit confusing. A vocabulary is an atomic collection of interrelated keywords that can be used in a JSON Schema. These keywords are not only for validation. i.e. you have previously documented annotation keywords like `title`, that come from the "Meta Data" vocabulary

Author: Okay, lets omit that

The `$vocabulary` keyword is used in meta-schemas to identify the vocabularies available for use in schemas described by that meta-schema. It is also used to indicate whether each vocabulary is required or optional, in the sense that an implementation must understand the required vocabularies in order to successfully process the schema.

* **Required and optional vocabularies:** If a vocabulary is required and an implementation does not recognize it, it must refuse to process any schemas that declare this meta-schema. If a vocabulary is optional, implementations that do not recognize it should proceed with processing such schemas.

> * **Mandatory:** The core vocabulary MUST always be included and set as required.

Reviewer: ```suggestion
* **Mandatory:** The Core vocabulary MUST always be included and set as required.
```

* **Non-inheritability:** Vocabularies defined in one meta-schema do not automatically apply to another meta-schema that references it. Each meta-schema must declare its vocabularies independently.

* **Recommendation:** Meta-schemas should always declare this keyword to clearly specify the vocabularies in use and avoid ambiguities.

## Examples

> {{<schema `'$vocabulary' for default meta-schema`>}}

Reviewer: ```suggestion
{{<schema `'$vocabulary' for default official 2020-12 meta-schema`>}}
```

{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://json-schema.org/draft/2020-12/schema",
  "$vocabulary": {
    "https://json-schema.org/draft/2020-12/vocab/core": true,
    "https://json-schema.org/draft/2020-12/vocab/applicator": true,
    "https://json-schema.org/draft/2020-12/vocab/unevaluated": true,
    "https://json-schema.org/draft/2020-12/vocab/validation": true,
    "https://json-schema.org/draft/2020-12/vocab/meta-data": true,
    "https://json-schema.org/draft/2020-12/vocab/format-annotation": true,
    "https://json-schema.org/draft/2020-12/vocab/content": true
  },
  "allOf" : [
    { "$ref": "meta/core" },
    { "$ref": "meta/applicator" },
    { "$ref": "meta/unevaluated" },
    { "$ref": "meta/validation" },
    { "$ref": "meta/meta-data" },
    { "$ref": "meta/format-annotation" },
    { "$ref": "meta/content" }
  ],
  // ...
}
{{</schema>}}

> {{<schema `Custom vocabulary with a custom keyword`>}}

Reviewer: This is not a custom vocabulary, but a vocabulary meta-schema

Author: can you provide an example of a custom vocabulary?

{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/meta/example-vocab",

>   "$dynamicAnchor": "meta",

Reviewer: I think you should explain `$dynamicAnchor: meta` in this page. You can explain it is set by convention to `meta` on the official meta-schemas, and it's a mechanism to enable meta-schema extensibility. By declaring `$dynamicAnchor: meta` here, you are essentially telling JSON Schema to validate EVERY subschema of the instance schema against the meta-schema, and not only the top level of it

Author: Should I add it under <learning-more> block or just below that schema block as a note?

Author: Should I add it under `leaning-more` block or just below that schema block as a note?

Reviewer: Yeah, that's a good point. Let's use the new `learning-more` block :)

  "type": [ "object", "boolean" ],
  "properties": {
    "minDate": {
      "type": "string",

>       "pattern": "\d\d\d\d-\d\d-\d\d",

Reviewer: This pattern is invalid JSON. You have to escape the backslashes

      "format": "date"
    }
  }
}
{{</schema>}}

> {{<schema `Meta-schema that includes the above vocabulary`>}}

Reviewer: Maybe add a variant of this example too importing the example vocabulary as optional? Just to exemplify that

  "$id": "https://example.com/meta/example-vocab",
  "$dynamicAnchor": "meta",
  "type": [ "object", "boolean" ],
  "properties": {
    "minDate": {
      "type": "string",
      "pattern": "\\d\\d\\d\\d-\\d\\d-\\d\\d",
      "format": "date"
    }
  }
}
{{</schema>}}

 * The `$dynamicAnchor: meta` declaration is set by convention to `meta` on the official meta-schemas. This setting serves as a mechanism to enable meta-schema extensibility. By declaring `$dynamicAnchor: meta` here, JSON Schema is configured to validate every subschema of the instance schema against the meta-schema, extending validation beyond just the top level.

{{<schema `Meta-schema with the above vocabulary as required`>}}
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/schema-required",
  "$dynamicAnchor": "meta",
  "$vocabulary": {
    "https://json-schema.org/draft/2020-12/vocab/core": true,
    "https://example.com/vocab/example-vocab": "true"
  },
  "allOf": [
    { "$ref": "https://json-schema.org/draft/2020-12/meta/core" },
    { "$ref": "https://example.com/meta/example-vocab" }
  ]
}
{{</schema>}}

> {{<schema `Meta-schema with the above vocabularyas optional`>}}

Reviewer: ```suggestion
{{<schema `Meta-schema with the above vocabulary as optional`>}}
```
